Auto generate test

brianhlee opened the pull request:
example is of a hosted zone

can add the rest once I get the LGTM

Commit:
initial commit

Commit:
example parameterize test with hosted zone particle

Commit:
added test ids and formatted the input data as a dictionary

Commit:
autogenerate test for particle classes

example is of a hosted zone

can add the rest once I get the good to go

davidyum commented:
Nice this is awesome lol
Going forward do we wanna put all the json for the tests in one json file or separate json out per particle?
Also curious if we have multiple particles does paramatize decorator make everything show up as one test?

Commit:
multiple moto tags

davidyum commented:
Can you make it so that update and delete tests can be optional. Some particles don't have update and moto3 delete is not implemented for some particles.

anovis commented:
Think about how we would handle those test cases since we still want tests for those. Maybe extra configs vars and then use another testing library   

Commit:
Merge remote-tracking branch 'upstream/develop' into auto-generate-test

Commit:
combined placebo mocking into the test generator

anovis commented on pcf/particle/aws/route53/hosted_zone.py:
```diff
@@ -41,12 +41,12 @@ class HostedZone(AWSResource):
 
     UNIQUE_KEYS = ["aws_resource.Name"]
 
-    def __init__(self, particle_definition):
```
can we add this to aws_resource so we dont have to edit every particle

brianhlee replied:
def __init__(self, particle_definition, session=None):
think you mean this one from cloudfront.py

it already is in aws_resource, but was never used from what I could tell. I think we need to edit every particle to have the session parameter, otherwise there would be no way to pass it in when creating the particle.

anovis replied:
yeah i think your right.  kinda of a pain but cant think of another way

anovis commented on pcf/test/particle/aws/test_particles.py:
```diff
@@ -0,0 +1,54 @@
+
+
+@pytest.mark.parametrize("definition,updated_definition,test_tag", values, ids=list(testdata.keys()))
+def test_apply(definition, updated_definition, test_tag):
```
test tag is confusion. test_type or something

anovis commented on pcf/test/particle/aws/test_particles.py:
```diff
@@ -0,0 +1,54 @@
+        particle.set_desired_state(State.running)
+        particle.apply()
+
+        assert particle.get_state() == State.running
```
i think we need to assert current def = input def.

anovis replied:
add            assert particle.is_state_definition_equivalent()

anovis commented on pcf/test/particle/aws/test_particles.py:
```diff
@@ -0,0 +1,54 @@
+            particle.set_desired_state(State.running)
+            particle.apply()
+
+            assert particle.is_state_definition_equivalent()
```
i think we need to assert current def = input def.

brianhlee replied:
isn't that what is_state_definition_equivalent() is doing?

anovis replied:
yeah your right. but we should that for the running state as well

anovis commented on pcf/test/particle/aws/test_particles.py:
```diff
@@ -0,0 +1,54 @@
+
+from pcf.particle.aws.route53 import hosted_zone
+from pcf.particle.aws.sqs.sqs_queue import SQSQueue
+from pcf.particle.aws.cloudfront.cloudfront import CloudFront
```
we will switch to use jacobs dynamic flavor importer so we dont need to add these

brianhlee replied:
do i just get rid of them, or do i need to set something up?

anovis replied:
we need to wait for his pr to get merged so just fyi. we can change in the next pr once we convert all the other tests

anovis commented on pcf/test/particle/aws/test_particles.py:
```diff
@@ -0,0 +1,54 @@
+from pcf.particle.aws.sqs.sqs_queue import SQSQueue
+from pcf.particle.aws.cloudfront.cloudfront import CloudFront
+
+with open("testdata.json") as f:
```
E   FileNotFoundError: [Errno 2] No such file or directory: 'testdata.json'

since pytest is run for different directory. use `os` to get the right location

Commit:
file path generalization and variable name change

anovis commented on pcf/test/particle/aws/testdata.json:
```diff
@@ -0,0 +1,167 @@
+      }
+    },
+    {
+      "pcf_name": "pcf_hosted_zone",
```
Can we change this so only the updated fields are here. Then you can just use the update util in the test function to get the full updated definition. This should reduce the file size a lot and make it much easier to read.

brianhlee replied:
made all the changes and pushed, but was wondering

what if a tag is removed? using update_dict still includes that removed tag
It's not a big deal, so I just pushed anyway, but wanted to make sure it wouldn't be an issue

Commit:
requested changes

assert state definition equivalency in the create apply
only include changed fields in the update dictionary

Commit:
fixed update_dict call
